fission/function.go

```diff
@@ -206,21 +231,47 @@ func fnUpdate(c *cli.Context) error {
 	})
 	checkErr(err, fmt.Sprintf("read function '%v'", fnName))
 
-	pkgName := function.Spec.Deployment.PackageRef.Name
+	var pkgName string
+	var isSrcPkg bool
+
+	// Since a deployment pkg may be built from the source pkg,
+	// we need to check whether source pkg is available or not.
+	if len(function.Spec.Source.PackageRef.Name) > 0 {
+		isSrcPkg = true
+		pkgName = function.Spec.Source.PackageRef.Name
+	} else {
+		isSrcPkg = false
+		pkgName = function.Spec.Deployment.PackageRef.Name
+	}
+
 	pkg, err := client.PackageGet(&api.ObjectMeta{
 		Name:      pkgName,
 		Namespace: api.NamespaceDefault,
 	})
 	checkErr(err, fmt.Sprintf("read package '%v'", pkgName))
 
 	envName := c.String("env")
-	fileName := c.String("code")
-	if len(fileName) == 0 {
-		fileName = c.String("package")
+	deployPkgName := c.String("code")
+	if len(deployPkgName) == 0 {
+		deployPkgName = c.String("package")
 	}
+	srcPkgName := c.String("srcpkg")
 
-	if len(envName) == 0 && len(fileName) == 0 {
-		fatal("Need --env or --code or --package argument.")
+	if len(envName) == 0 && len(deployPkgName) == 0 && len(srcPkgName) == 0 {
+		fatal("Need --env or --code or --package or --srcpkg argument.")
+	}
+
+	// We should not allow a user to update deployment pkg if the source
+	// pkg is not empty, since the deployment pkg may be overwritten with
+	// a new one, which is built from the old source pkg, once env builder
+	// finished it's build process.
+	if isSrcPkg && len(deployPkgName) > 0 && len(srcPkgName) == 0 {
```

Comment:
How about allowing this, but choosing not to do a build. In other words, do a build only if a function has a source but no deployment package.